File: attocode/src/integrations/streaming/pty-shell.ts
/**
 * Persistent PTY Shell
 *
 * Provides a persistent shell session that maintains state between commands.
 * Uses Node.js child_process with PTY-like behavior.
 *
 * Key features:
 * - Persistent shell process (bash/zsh)
 * - Command history tracking
 * - Environment inheritance
 * - Working directory state
 * - Output buffering with timeout
 */

import { spawn, ChildProcess } from 'node:child_process';
import { platform } from 'node:os';

// =============================================================================
// TYPES
// =============================================================================

export interface PTYShellConfig {
  /** Shell to use (default: auto-detect) */
  shell?: string;
  /** Initial working directory */
  cwd?: string;
  /** Environment variables */
  env?: Record<string, string>;
  /** Command timeout in ms (default: 30000) */
  timeout?: number;
  /** Maximum output buffer size (default: 1MB) */
  maxOutputSize?: number;
  /** Shell prompt to detect command completion */
  promptPattern?: string;
}

export interface CommandResult {
  output: string;
  exitCode: number | null;
  duration: number;
  timedOut: boolean;
}

export interface ShellState {
  cwd: string;
  env: Record<string, string>;
  history: string[];
  isRunning: boolean;
  pid?: number;
}

export type PTYEvent =
  | { type: 'shell.started'; pid: number; shell: string }
  | { type: 'shell.stopped'; exitCode: number | null }
  | { type: 'command.start'; command: string }
  | { type: 'command.output'; data: string }
  | { type: 'command.complete'; result: CommandResult }
  | { type: 'command.timeout'; command: string }
  | { type: 'error'; error: string };

export type PTYEventListener = (event: PTYEvent) => void;

// =============================================================================
// PTY SHELL MANAGER
// =============================================================================

/**
 * Manages persistent shell sessions.
 */
export class PTYShellManager {
  private config: Required<PTYShellConfig>;
  private process: ChildProcess | null = null;
  private outputBuffer: string = '';
  private commandHistory: string[] = [];
  private currentCwd: string;
  private currentEnv: Record<string, string>;
  private commandResolve: ((result: CommandResult) => void) | null = null;
  private commandStartTime: number = 0;
  private commandTimeoutId: NodeJS.Timeout | null = null;
  private eventListeners: Set<PTYEventListener> = new Set();
  private readonly endMarker: string;

  constructor(config: PTYShellConfig = {}) {
    this.config = {
      shell: config.shell || this.detectShell(),
      cwd: config.cwd || process.cwd(),
      env: config.env || {},
      timeout: config.timeout ?? 30000,
      maxOutputSize: config.maxOutputSize ?? 1024 * 1024, // 1MB
      promptPattern: config.promptPattern || '__CMD_DONE__',
    };
    this.currentCwd = this.config.cwd;
    this.currentEnv = { ...process.env, ...this.config.env } as Record<string, string>;
    this.endMarker = `\necho "${this.config.promptPattern} $?"`;
  }

  /**
   * Auto-detect the user's preferred shell.
   */
  private detectShell(): string {
    const userShell = process.env.SHELL;
    if (userShell) return userShell;

    if (platform() === 'win32') {
      return 'cmd.exe';
    }

    return '/bin/bash';
  }

  /**
   * Start the persistent shell.
   */
  async start(): Promise<void> {
    if (this.process && !this.process.killed) {
      return; // Already running
    }

    return new Promise((resolve, reject) => {
      const env = {
        ...process.env,
        ...this.config.env,
        // Disable shell prompt customization that might interfere
        PS1: '$ ',
        PROMPT_COMMAND: '',
      };

      this.process = spawn(this.config.shell, [], {
        cwd: this.config.cwd,
        env,
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: false, // We're spawning a shell directly
      });

      // Set up output handlers
      this.process.stdout?.on('data', (data: Buffer) => {
        const output = data.toString();
        this.handleOutput(output);
      });

      this.process.stderr?.on('data', (data: Buffer) => {
        const output = data.toString();
        this.handleOutput(output);
      });

      this.process.on('error', (error) => {
        this.emitEvent({ type: 'error', error: error.message });
        reject(error);
      });

      this.process.on('exit', (code) => {
        this.emitEvent({ type: 'shell.stopped', exitCode: code });
        this.process = null;
      });

      // Wait a bit for shell to initialize
      setTimeout(() => {
        if (this.process && this.process.pid) {
          this.emitEvent({
            type: 'shell.started',
            pid: this.process.pid,
            shell: this.config.shell,
          });
          resolve();
        } else {
          reject(new Error('Failed to start shell'));
        }
      }, 100);
    });
  }

  /**
   * Handle output from the shell.
   */
  private handleOutput(data: string): void {
    this.outputBuffer += data;
    this.emitEvent({ type: 'command.output', data });

    // Trim buffer if too large
    if (this.outputBuffer.length > this.config.maxOutputSize) {
      this.outputBuffer = this.outputBuffer.slice(-this.config.maxOutputSize);
    }

    // Check for command completion
    if (this.commandResolve && this.outputBuffer.includes(this.config.promptPattern)) {
      this.handleCommandComplete();
    }
  }

  /**
   * Handle command completion.
   */
  private handleCommandComplete(): void {
    if (!this.commandResolve) return;

    // Clear timeout
    if (this.commandTimeoutId) {
      clearTimeout(this.commandTimeoutId);
      this.commandTimeoutId = null;
    }

    // Extract output and exit code
    const markerIndex = this.outputBuffer.lastIndexOf(this.config.promptPattern);
    let output = this.outputBuffer.slice(0, markerIndex).trim();
    const exitCodeMatch = this.outputBuffer.slice(markerIndex).match(/(\d+)/);
    const exitCode = exitCodeMatch ? parseInt(exitCodeMatch[1], 10) : null;

    // Remove the echo command from output
    const echoIndex = output.lastIndexOf('echo "');
    if (echoIndex !== -1) {
      output = output.slice(0, echoIndex).trim();
    }

    // Also remove any command echo (bash echoes input commands)
    const lines = output.split('\n');
    if (lines.length > 1 && this.commandHistory.length > 0) {
      const lastCmd = this.commandHistory[this.commandHistory.length - 1];
      if (lines[0].endsWith(lastCmd) || lines[0].includes(lastCmd)) {
        lines.shift();
        output = lines.join('\n');
      }
    }

    const result: CommandResult = {
      output: output.trim(),
      exitCode,
      duration: Date.now() - this.commandStartTime,
      timedOut: false,
    };

    this.emitEvent({ type: 'command.complete', result });

    const resolve = this.commandResolve;
    this.commandResolve = null;
    this.outputBuffer = '';

    resolve(result);
  }

  /**
   * Execute a command in the persistent shell.
   */
  async execute(command: string): Promise<CommandResult> {
    if (!this.process || this.process.killed) {
      await this.start();
    }

    if (!this.process?.stdin) {
      throw new Error('Shell not running');
    }

    // Track command
    this.commandHistory.push(command);
    this.commandStartTime = Date.now();
    this.outputBuffer = '';

    this.emitEvent({ type: 'command.start', command });

    return new Promise((resolve) => {
      this.commandResolve = resolve;

      // Set timeout
      this.commandTimeoutId = setTimeout(() => {
        if (this.commandResolve) {
          this.emitEvent({ type: 'command.timeout', command });

          const result: CommandResult = {
            output: this.outputBuffer.trim(),
            exitCode: null,
            duration: Date.now() - this.commandStartTime,
            timedOut: true,
          };

          this.commandResolve = null;
          this.outputBuffer = '';

          resolve(result);
        }
      }, this.config.timeout);

      // Write command with completion marker
      const fullCommand = `${command}${this.endMarker}\n`;
      this.process!.stdin!.write(fullCommand);
    });
  }

  /**
   * Change the working directory.
   */
  async cd(directory: string): Promise<CommandResult> {
    const result = await this.execute(`cd "${directory}" && pwd`);
    if (result.exitCode === 0) {
      this.currentCwd = result.output.trim();
    }
    return result;
  }

  /**
   * Set an environment variable.
   */
  async setEnv(key: string, value: string): Promise<void> {
    await this.execute(`export ${key}="${value}"`);
    this.currentEnv[key] = value;
  }

  /**
   * Get the current state.
   */
  getState(): ShellState {
    return {
      cwd: this.currentCwd,
      env: this.currentEnv,
      history: [...this.commandHistory],
      isRunning: this.process !== null && !this.process.killed,
      pid: this.process?.pid,
    };
  }

  /**
   * Get command history.
   */
  getHistory(): string[] {
    return [...this.commandHistory];
  }

  /**
   * Clear command history.
   */
  clearHistory(): void {
    this.commandHistory = [];
  }

  /**
   * Subscribe to events.
   */
  subscribe(listener: PTYEventListener): () => void {
    this.eventListeners.add(listener);
    return () => this.eventListeners.delete(listener);
  }

  /**
   * Emit an event.
   */
  private emitEvent(event: PTYEvent): void {
    for (const listener of this.eventListeners) {
      try {
        listener(event);
      } catch {
        // Ignore listener errors
      }
    }
  }

  /**
   * Stop the shell.
   */
  async stop(): Promise<void> {
    if (!this.process) return;

    return new Promise((resolve) => {
      if (!this.process) {
        resolve();
        return;
      }

      const timeout = setTimeout(() => {
        this.process?.kill('SIGKILL');
        resolve();
      }, 1000);

      this.process.once('exit', () => {
        clearTimeout(timeout);
        resolve();
      });

      // Try graceful exit first
      this.process.stdin?.write('exit\n');
    });
  }

  /**
   * Cleanup resources.
   */
  async cleanup(): Promise<void> {
    if (this.commandTimeoutId) {
      clearTimeout(this.commandTimeoutId);
    }
    await this.stop();
    this.eventListeners.clear();
  }
}

// =============================================================================
// FACTORY FUNCTIONS
// =============================================================================

/**
 * Create a PTY shell manager.
 */
export function createPTYShell(config?: PTYShellConfig): PTYShellManager {
  return new PTYShellManager(config);
}

/**
 * Create and start a PTY shell.
 */
export async function createAndStartPTYShell(config?: PTYShellConfig): Promise<PTYShellManager> {
  const shell = new PTYShellManager(config);
  await shell.start();
  return shell;
}

// =============================================================================
// TOOL FACTORY
// =============================================================================

/**
 * Create a shell execution tool that uses the persistent PTY.
 */
export function createPTYShellTool(shell: PTYShellManager) {
  return {
    name: 'shell_exec',
    description:
      'Execute a command in the persistent shell session. Maintains state between calls (working directory, environment variables, etc.)',
    parameters: {
      type: 'object',
      properties: {
        command: {
          type: 'string',
          description: 'The command to execute',
        },
      },
      required: ['command'],
    },
    dangerLevel: 'dangerous' as const,
    async execute(args: Record<string, unknown>) {
      const command = args.command as string;
      if (!command) {
        return 'Error: command is required';
      }

      const result = await shell.execute(command);

      if (result.timedOut) {
        return `Command timed out after ${result.duration}ms. Partial output:\n${result.output}`;
      }

      const exitInfo = result.exitCode !== 0 ? ` (exit code: ${result.exitCode})` : '';
      return `${result.output}${exitInfo}`;
    },
  };
}

/**
 * Format shell state for display.
 */
export function formatShellState(state: ShellState): string {
  const lines = [
    `Shell: ${state.isRunning ? 'Running' : 'Stopped'}${state.pid ? ` (PID: ${state.pid})` : ''}`,
    `CWD: ${state.cwd}`,
    `History: ${state.history.length} commands`,
  ];

  if (state.history.length > 0) {
    lines.push('Recent commands:');
    for (const cmd of state.history.slice(-5)) {
      lines.push(`  $ ${cmd}`);
    }
  }

  return lines.join('\n');
}

File: attocode/attocode_py/src/attocode/integrations/streaming/pty_shell.py
"""Persistent PTY shell.

Provides a persistent shell session that maintains state between
commands (working directory, environment, etc.) using asyncio
subprocess management.
"""

from __future__ import annotations

import asyncio
import os
import platform
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any


# =============================================================================
# Types
# =============================================================================


@dataclass
class PTYShellConfig:
    """PTY shell configuration."""

    shell: str = ""  # Auto-detect if empty
    cwd: str = ""  # Default to cwd if empty
    env: dict[str, str] = field(default_factory=dict)
    timeout: float = 30.0  # Seconds
    max_output_size: int = 1_048_576  # 1MB
    prompt_pattern: str = "__CMD_DONE__"


@dataclass
class CommandResult:
    """Result of a shell command."""

    output: str
    exit_code: int | None
    duration: float  # Seconds
    timed_out: bool


@dataclass
class ShellState:
    """Current state of the shell."""

    cwd: str
    env: dict[str, str]
    history: list[str]
    is_running: bool
    pid: int | None = None


PTYEventListener = Callable[[str, dict[str, Any]], None]


# =============================================================================
# PTY Shell Manager
# =============================================================================


class PTYShellManager:
    """Manages a persistent shell session.

    Commands run in sequence in the same shell process,
    preserving environment and working directory between calls.
    Uses a sentinel marker to detect command completion.
    """

    def __init__(self, config: PTYShellConfig | None = None) -> None:
        cfg = config or PTYShellConfig()
        self._shell = cfg.shell or self._detect_shell()
        self._cwd = cfg.cwd or os.getcwd()
        self._base_env = {**os.environ, **cfg.env}
        self._timeout = cfg.timeout
        self._max_output_size = cfg.max_output_size
        self._prompt_pattern = cfg.prompt_pattern

        self._process: asyncio.subprocess.Process | None = None
        self._history: list[str] = []
        self._listeners: set[PTYEventListener] = set()

    @staticmethod
    def _detect_shell() -> str:
        """Auto-detect the user's shell."""
        user_shell = os.environ.get("SHELL")
        if user_shell:
            return user_shell
        if platform.system() == "Windows":
            return "cmd.exe"
        return "/bin/bash"

    async def start(self) -> None:
        """Start the persistent shell process."""
        if self._process is not None and self._process.returncode is None:
            return  # Already running

        env = {**self._base_env, "PS1": "$ ", "PROMPT_COMMAND": ""}

        self._process = await asyncio.create_subprocess_shell(
            self._shell,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
            cwd=self._cwd,
            env=env,
        )

        # Give shell time to initialize
        await asyncio.sleep(0.1)

        if self._process.pid:
            self._emit("shell.started", {
                "pid": self._process.pid,
                "shell": self._shell,
            })
        else:
            raise RuntimeError("Failed to start shell")

    async def run_command(self, command: str) -> CommandResult:
        """Run a command in the persistent shell.

        Uses a sentinel marker pattern to detect when the command
        has completed and to capture the exit code.
        """
        if self._process is None or self._process.returncode is not None:
            await self.start()

        assert self._process is not None
        assert self._process.stdin is not None
        assert self._process.stdout is not None

        self._history.append(command)
        start_time = time.monotonic()
        self._emit("command.start", {"command": command})

        # Write command with end marker that echoes the exit code
        end_marker = f'\necho "{self._prompt_pattern} $?"\n'
        full_command = command + end_marker
        self._process.stdin.write(full_command.encode())
        await self._process.stdin.drain()

        # Read output until we see the marker
        output_parts: list[str] = []
        timed_out = False

        try:
            output = await asyncio.wait_for(
                self._read_until_marker(output_parts),
                timeout=self._timeout,
            )
        except asyncio.TimeoutError:
            timed_out = True
            output = "".join(output_parts)
            self._emit("command.timeout", {"command": command})

        duration = time.monotonic() - start_time

        # Parse exit code from marker line
        exit_code: int | None = None
        if not timed_out and self._prompt_pattern in output:
            marker_idx = output.rfind(self._prompt_pattern)
            after_marker = output[marker_idx + len(self._prompt_pattern):]
            parts = after_marker.strip().split()
            digits = "".join(c for c in parts[0] if c.isdigit()) if parts else ""
            exit_code = int(digits) if digits else None
            output = output[:marker_idx]

        # Clean up echo command from output
        lines = output.split("\n")
        cleaned: list[str] = []
        for line in lines:
            if f'echo "{self._prompt_pattern}' in line:
                continue
            cleaned.append(line)
        output = "\n".join(cleaned).strip()

        # Trim if too large
        if len(output) > self._max_output_size:
            output = output[-self._max_output_size:]

        result = CommandResult(
            output=output,
            exit_code=exit_code,
            duration=duration,
            timed_out=timed_out,
        )
        self._emit("command.complete", {"result": result})
        return result

    async def _read_until_marker(self, parts: list[str]) -> str:
        """Read from stdout until the prompt pattern is found."""
        assert self._process is not None
        assert self._process.stdout is not None

        buffer = ""
        while True:
            data = await self._process.stdout.read(4096)
            if not data:
                break
            chunk = data.decode(errors="replace")
            parts.append(chunk)
            buffer += chunk
            if self._prompt_pattern in buffer:
                break
        return buffer

    async def cd(self, directory: str) -> CommandResult:
        """Change working directory."""
        result = await self.run_command(f'cd "{directory}" && pwd')
        if result.exit_code == 0:
            self._cwd = result.output.strip()
        return result

    async def set_env(self, key: str, value: str) -> None:
        """Set an environment variable in the shell."""
        await self.run_command(f'export {key}="{value}"')
        self._base_env[key] = value

    def get_state(self) -> ShellState:
        """Get current shell state."""
        return ShellState(
            cwd=self._cwd,
            env=dict(self._base_env),
            history=list(self._history),
            is_running=(
                self._process is not None
                and self._process.returncode is None
            ),
            pid=self._process.pid if self._process else None,
        )

    def get_history(self) -> list[str]:
        """Get command history."""
        return list(self._history)

    def clear_history(self) -> None:
        """Clear command history."""
        self._history.clear()

    def on(self, listener: PTYEventListener) -> Callable[[], None]:
        """Subscribe to PTY events. Returns unsubscribe function."""
        self._listeners.add(listener)
        return lambda: self._listeners.discard(listener)

    async def stop(self) -> None:
        """Stop the shell process."""
        if self._process is None:
            return

        if self._process.returncode is None:
            # Try graceful exit
            if self._process.stdin:
                try:
                    self._process.stdin.write(b"exit\n")
                    await self._process.stdin.drain()
                except (BrokenPipeError, ConnectionResetError):
                    pass

            try:
                await asyncio.wait_for(self._process.wait(), timeout=1.0)
            except asyncio.TimeoutError:
                self._process.kill()
                await self._process.wait()

        exit_code = self._process.returncode
        self._process = None
        self._emit("shell.stopped", {"exit_code": exit_code})

    async def cleanup(self) -> None:
        """Stop shell and clear listeners."""
        await self.stop()
        self._listeners.clear()

    def _emit(self, event: str, data: dict[str, Any]) -> None:
        for listener in self._listeners:
            try:
                listener(event, data)
            except Exception:
                pass


def format_shell_state(state: ShellState) -> str:
    """Format shell state for display."""
    status = "Running" if state.is_running else "Stopped"
    pid_str = f" (PID: {state.pid})" if state.pid else ""
    lines = [
        f"Shell: {status}{pid_str}",
        f"CWD: {state.cwd}",
        f"History: {len(state.history)} commands",
    ]
    if state.history:
        lines.append("Recent commands:")
        for cmd in state.history[-5:]:
            lines.append(f"  $ {cmd}")
    return "\n".join(lines)
